addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/connector/DSMRSerialConnector.java

```diff
@@ -0,0 +1,282 @@
+ * <p>
+ * This class will handle communication with the Serial Port itself and notify listeners about events and received P1
+ * telegrams.
+ *
+ * @author M. Volaart - Initial contribution
+ * @author Hilbrand Bouwkamp - Split code and moved DSMR specific handling to separate class.
+ */
+@NonNullByDefault
+public class DSMRSerialConnector extends DSMRBaseConnector implements SerialPortEventListener {
+    /**
+     * Serial port read time out in milliseconds.
+     */
+    private static final int SERIAL_PORT_READ_TIMEOUT_MILLISECONDS = (int) TimeUnit.SECONDS.toMillis(15);
+
+    /**
+     * The receive threshold/time out set on the serial port.
+     */
+    private static final int SERIAL_TIMEOUT_MILLISECONDS = 1000;
+
+    private final Logger logger = LoggerFactory.getLogger(DSMRSerialConnector.class);
+
+    /**
+     * Manager to get new port from.
+     */
+    private final SerialPortManager portManager;
+
+    /**
+     * Serial port name.
+     */
+    private final String serialPortName;
+
+    /**
+     * Serial port instance.
+     */
+    @Nullable
+    private SerialPort serialPort;
+
+    /**
+     * DSMR Connector listener.
+     */
+    private final DSMRConnectorListener dsmrConnectorListener;
+
+    /**
+     * The portLock is used for the shared data used when opening and closing
+     * the port. The following shared data must be guarded by the lock:
+     * SerialPort, BufferedReader
+     */
+    private final Object portLock = new Object();
+
+    /**
+     * Creates a new DSMR serial connector. This is only a reference to a port. The port will
+     * not be opened nor it is checked if the DSMR Port can successfully be opened.
+     *
+     * @param portManager           Serial Port Manager
+     * @param serialPortName        Device identifier of the port (e.g. /dev/ttyUSB0)
+     * @param dsmrConnectorListener The listener to send error or received data from the port
+     *
+     */
+    public DSMRSerialConnector(SerialPortManager portManager, String serialPortName,
+            DSMRConnectorListener dsmrConnectorListener) {
+        super(dsmrConnectorListener);
+        this.portManager = portManager;
+        this.serialPortName = serialPortName;
+        this.dsmrConnectorListener = dsmrConnectorListener;
+    }
+
+    public String getPortName() {
+        return serialPortName;
+    }
+
+    /**
+     * Opens the Operation System Serial Port.
+     *
+     * @param portSettings The serial port settings to open the port with
+     */
+    public void open(DSMRSerialSettings portSettings) {
+        DSMRConnectorErrorEvent errorEvent = null;
+
+        synchronized (portLock) {
+            try {
+                logger.trace("Opening port {}", serialPortName);
+                // Opening Operating System Serial Port
+                CommPortIdentifier portIdentifier = portManager.getIdentifier(serialPortName);
+                serialPort = portIdentifier.open(DSMRBindingConstants.DSMR_PORT_NAME,
```

Comment:
To fix all these potential null pointer exceptions you need to assign `serialPort` to a local variable. Then do a null check and handle possible null values.

The reason for this is that the value of the `serialPort` property could be changed by another thread at any moment in time. So you will never know if `serialPort` is still not `null` after doing a null check. :-)

Can you also fix the other potential NPEs? Eclipse shows there are a total of 15 with 12 in this method.

Reply:
O yes those null pointers warnings.... Maven shows me even more: 36! And these feel kind of sad. Doing an initialization of a field and in the next statement call a method on that field will result in a null pointer warning (although Eclipse doesn't warn me here, but maven does):
> dsmrDeviceThread = new Thread(dsmrDeviceRunnable);
> dsmrDeviceThread.start();   // <-- null pointer warning

Even wrapping code with a null check doesn't satisfy. I know theoretically they can be null, but given the working of the smart home framework it will not be called in a multi threading manner.  The tool-satisfying workaround is to introduce a local variable. Even if it theoretically does happen and I introduce a local variable. Do I really want to call start in case in between the initialization and calling start the field was assigned null? If I would introduce a local variable here I would have started a dangling thread that will never be closed since there is no reference. So this means to really fix it I would have to synchronize the call and the null assignment to that field to make sure it will work. That basically means all code must be made thread safe. But is that really needed?

Regarding the serialPort: it can never be null since all calls are wrapped in synchronized blocks,. Thus the warnings are all false-positives. The sad part: Introducing local variables, but removing the synchronized blocks would totally satisfy the tool, but also totally break the code...
**The current checking of @NonNullByDefault adds an implicit `ThreadSafe` to the class** and I'm not sure if that is really desired.

So I'm not sure what the best option is: Introduce local variables and possible cause really hard to find bugs because fields are potential assigned null and after that operations are performed on local variables or make all classes also threadsafe or simply remove the @NonNullByDefault?
Working around tooling always bothers me. I prefer to live with it's flaws.
So in the mean time I'll read them as to speak with my favorite line from Pirates of the caribbean:
> “The code is more what you’d call ‘guidelines’ than actual rules.” – Hector Barbossa

Reply:
Yes I agree the tooling could be a bit smarter at times. I think you're also allowed to use `@SuppressWarnings("null")` for suppressing real false positives.